theano/gof/utils.py

```diff
@@ -210,6 +210,19 @@ def __str__(self):
 
         return type.__new__(cls, name, bases, dct)
 
+    _cache = {}
+
+    def __call__(cls, *args, **kwargs):
+        props = cls.__dict__.get('__props__', None)
+        if props is None:
+            return super(MetaObject, cls).__call__(*args, **kwargs)
+        key = (cls, args, tuple(kwargs.values()))
```

Comment:
For the problem with dict as parameter, find where we pass a dict and modify that code to pass a frozendict instead.

Maybe around tensor/opt.py line 349, you need something like:
inplace_patern = theano.misc.fronzendict.frozendict(inplace_pattern)